flatten_json.go

```diff
@@ -208,42 +227,56 @@ func (fj *flattenJSON) readObject(pathName []byte) error {
 				val, alt, err = fj.readNumber()
 				isLeaf = true
 			case '[':
-				if !fj.tracker.IsNameUsed(memberName) {
+				if !pathNode.IsNameUsed(memberName) {
 					fj.skipping++
 				}
-				var arrayPath []byte
-				if fj.skipping == 0 {
-					arrayPath = pathForChild(pathName, memberName)
-				}
 
 				if fj.skipping > 0 || !memberIsUsed {
 					err = fj.skipBlock('[', ']')
 				} else {
-					err = fj.readArray(arrayPath)
+					arrayPathNode, exists := pathNode.Get(memberName)
+					if !exists {
+						// Arrays are interesting, they can be field or node.
+						// Given this case:
+						//  { "geo": { "coords": [{"coordinates": [1,2,3]}] } }
+						// "coords" is a node.
+						// "coordinates" is a field.
+						arrayPathNode = pathNode
+					}
+
+					err = fj.readArray(pathNode.PathForMemberName(memberName), arrayPathNode)
 				}
 				if err != nil {
 					return err
 				}
-				if !fj.tracker.IsNameUsed(memberName) {
+				if !pathNode.IsNameUsed(memberName) {
 					fj.skipping--
 				}
 			case '{':
-				if !fj.tracker.IsNameUsed(memberName) {
+				if !pathNode.IsNameUsed(memberName) {
 					fj.skipping++
 				}
-				var objectPath []byte
-				if fj.skipping == 0 {
-					objectPath = pathForChild(pathName, memberName)
-				}
 				if fj.skipping > 0 || !memberIsUsed {
 					err = fj.skipBlock('{', '}')
 				} else {
-					err = fj.readObject(objectPath)
+					objectPathNode, exists := pathNode.Get(memberName)
```

Comment:
would `ok` be a bit more idiomatic than `exists`? (Not a rhetorical question, maybe lots of people use things other than `ok` and I just haven't seen it.)

Reply:
`exists` is a problem of mine :) I am used for using `exists` as variable name since it makes more sense to me, but I know the go community went for `ok` - so I'll align this code and use `ok` (even though it's not natural for me, it's just better to align with conventions).

Reply:
Renamed to "ok"